src/utils.ts

```diff
@@ -303,3 +313,49 @@ export async function createUnsignedFillRelayTransaction(
     { from: simulatedRelayerAddress }
   );
 }
+
+/**
+ * Search back in time for the first block at or older than the desired lookback.
+ *
+ * The approach here is as below:
+ * 1. We use the skip distances, which are meant to be a couple of hours or less for each chain.
+ * 2. Fetch the first block one skip distance from the latest block. Check its timestamp to see how far back it is
+ * relatively to the latest block.
+ * 3. Apply a multiplier to the skip distance based on the time difference and find a block with the desired lookback
+ * This assumes the block production speed is roughly constant.
+ * 4. Check the timestamp again, if we still haven't achieved the desired lookback, repeat step 3 and go back further.
+ * Otherwise, return the block number.
+ *
+ * This is only a rough estimate and can go back further than desired. However, this would minimize the number
+ * of requests and is generally accurate unless the rate of block production has been very volatile over a
+ * short period of time.
+ *
+ * @param provider A valid provider - will be used to find the block with desired lookback.
+ * @param desiredLookback Desired lookback in seconds (e.g. 86400 seconds or 1 day).
+ * @return The block number that's at or older than the desired lookback.
+ */
+export async function findBlockAtOrOlder(
+  provider: providers.Provider | L2Provider<providers.Provider>,
+  desiredLookback: number
+): Promise<number> {
+  const latestBlock = await provider.getBlock("latest");
+  const latestBlockTimestamp = latestBlock.timestamp;
+  const desiredTimestamp = latestBlockTimestamp - desiredLookback;
+  const chainId = (await provider.getNetwork()).chainId;
+  let lastSkipDistance = BlockScanSkipDistances[chainId];
+  let lastBlockNumber = latestBlock.number;
```

Comment:
Why not make `latestBlockTimestamp` a variable instead of a constant instead of setting both `lastBlockNumber` and `latestBlockTimestamp` equal to `latestBlock.number`

I think overall try to condense the number of variables and re-use them in this function where possible to make it more readable

Reply:
Great suggestion! Updated.

Reply:
You could set the variables like this and then remove the section in your `while` loop that "skips the first base case where the last looked up block and the latest block are the same"

```ts
let toBlock = await provider.getBlock("latest");
let toBlockNumber = toBlock.number
let toBlockTimestamp = toBlock.timestamp;
const desiredTimestamp = toBlockTimestamp - desiredLookback;
let skipDistance = BlockScanSkipDistances[chainId];
let fromBlockNumber = toBlockNumber - skipDistance;
let fromBlock = await provider.getBlock(fromBlockNumber);
let fromBlockTimestamp = fromBlock.timestamp

while(fromBlockTimestamp > desiredTimestamp) {
   const blockSpeed = skipDistance / (toBlockTimestamp - fromBlockTimestamp);
   skipDistance = Math.floor(blockSpeed * (fromBlockTimestamp - desiredTimestamp));
   fromBlockNumber -= skipDistance;
   const block = await provider.getBlock(fromBlockNumber);
   // Set toBlock equal to current fromBlock and then decrement fromBlock
   toBlockTimestamp = fromBlockTimestamp
   fromBlockTimestamp = block.timestamp
}
```

Reply:
This makes sense. Love the variable names too! Updated.